### PR TITLE
Feature/jenkins 35789 35794 stop and rerun builds

### PR DESCRIPTION
**Pull request by cliffmeyers**
# Description
This PR includes includes a lot of work that had queued up while waiting for #392 to be merged. It also contains quite a bit of code already seen in #440. In short, this PR covers: refactoring the "RunPipeline" component to support Start and Stop (and moving it to core-js so it can be shared across dashboard and personalization), stopping a build from Activity tab and Run Details, re-running a failed build Activity tab and Run Details.
- See [JENKINS-36687](https://issues.jenkins-ci.org/browse/JENKINS-36687)
- See [JENKINS-35796](https://issues.jenkins-ci.org/browse/JENKINS-35796)
- See [JENKINS-36974](https://issues.jenkins-ci.org/browse/JENKINS-36974)
- See [JENKINS-35794](https://issues.jenkins-ci.org/browse/JENKINS-35794)
- See [JENKINS-35789](https://issues.jenkins-ci.org/browse/JENKINS-35789)
- Known Issues:
   - Cannot stop items that are queued; this is covered by ticket: [JENKINS-37835](https://issues.jenkins-ci.org/browse/JENKINS-37835)
- TODO:
   - [x] Fix a bug where the Replay button shows for all job types. It should only show for regular Pipeline and Multibranch Pipeline
   - [x] Need more unit tests for UrlBuilder.buildRunDetailsUrlFromQueue
   - [x] ~~Find a better way to prevent duplicate toasts being fired when a job is started or stopped. This is handled now in ToastService - and while it works - it's perhaps concerning that multiple instances of Run button bound to the same pipeline are firing dup toasts, and that Favorites from the dashboard will also attempt to propagate a toast.~~  Will be addressed in: https://issues.jenkins-ci.org/browse/JENKINS-37425

# Submitter checklist
- [x] Link to JIRA ticket in description, if appropriate.
- [x] Change is code complete and matches issue description
- [x] Appropriate unit or acceptance tests or explanation to why this change has no tests
- [x] Reviewer's manual test instructions provided in PR description. See Reviewer's first task below.
- [x] Ran Acceptance Test Harness against PR changes.

# Reviewer checklist
- [ ] Reviewed the code
- [ ] Verified that the appropriate tests have been written or valid explanation given
- [ ] Run the changes and verified the change matches the issue description

Ensure you have a freestyle, pipeline and multi-branch pipeline job available, ideally nested in some sub-folders and with git-flow style branch names to test for everyone's favorite, URI encoding bugs.

For testing the refactor of existing Run functionality:
- [ ] Navigate to Activity tab for a non multi-branch project. Test that the "Run" button here works as expected. The "Run" button should not change state after clicked, allowing multiple runs. Ensure the "Open" link displays Run Details correctly.
- [ ] Navigate to the Branches tab for a multi-branch project. Test that each branch can be run. Ensure the "Open" link displays Run Details correctly.
- [ ] Do the same for PR's tab.

For testing "Stop" button:
- [ ] Navigate to Activity tab for a non-multi branch project and press "Run" button. Verify that the new run displays a Stop button and that it works.
- [ ] Navigate to the Branches tab for a multi-branch project, press "Run" for a branch and verify that pressing "Stop" stops the run.
- [ ] For any in-progress run, click it to navigate to Run Details. Press the "Stop" button and verify the build stops and UI updates.

For "testing Re-run", each project will need at least one failed or aborted build, or you can use the new-ish Run / Stop functionality to create aborted builds:
- [ ] Navigate to Activity tab for each project. Any failed or aborted build should include a Re-run button for pipeline and multi-branch but *not* for freestyle.
- [ ] Clicking Re-run button should launch a new build. Verify clicking "Open" in the toast opens the Run Details screen correctly.
- [ ] Navigate to Run Details for a failed or aborted build. It should include a "Re-run" button to the right of the tabs.
- [ ] Clicking re-run should launch a new build, and update the screen / URL to display the updated Run Details screen.

@jenkinsci/code-reviewers @reviewbybees

### CODE REVIEW

**Comment by i386** [comment]
🐛  favourite stars are missing :(

jenkins blue ocean 2016-08-31 08-46-09

**Comment by i386** [comment]
@cliffmeyers I tested stop on freestyle and Pipeline at it LGTM 👍

**Comment by i386** [comment]
🐛 cannot stop a FreeStyle job in the queue

**Comment by i386** [comment]
🐛 stopping a Pipeline from the activity results in a console error

```
blueocean.js:795 Uncaught TypeError: Cannot read property 'self' of undefinedstopRun @ blueocean.js:795_onStopClick @ blueocean.js:578onClick @ blueocean.js:626invokeGuardedCallback @ blueocean.js:38820executeDispatch @ blueocean.js:31421executeDispatchesInOrder @ blueocean.js:31441executeDispatchesAndRelease @ blueocean.js:30920executeDispatchesAndReleaseTopLevel @ blueocean.js:30931forEachAccumulated @ blueocean.js:44523processEventQueue @ blueocean.js:31093runEventQueueInBatch @ blueocean.js:38878handleTopLevel @ blueocean.js:38889handleTopLevelImpl @ blueocean.js:38956perform @ blueocean.js:44099batchedUpdates @ blueocean.js:38138batchedUpdates @ blueocean.js:42009dispatchEvent @ blueocean.js:39033
```

**Comment by cliffmeyers** [comment]
Ticket created :)  https://issues.jenkins-ci.org/browse/JENKINS-37834

**Comment by i386** [comment]
Removing items in the queue can be covered in https://issues.jenkins-ci.org/browse/JENKINS-37835

Happy for this to get merged when you've got 🐝's - behaviour seems to checkout 👍

**Comment by michaelneale** [comment]
this needs to be updated to master, not sure if this should stay open until there is ability to stop queued things (probably yes...)

**Comment by cliffmeyers** [comment]
I think this is ready for code review. Spending a little time to clean up some of the Toast stuff, but the code changes should be major for that I don't think.

File: blueocean-core-js/src/js/ToastService.js [hunk 1]
```diff
@@ -25,16 +25,24 @@ export class ToastService {
      * @returns {number} unique ID of toast
      */
     newToast(toast) {
-        if (!toast.id) {
-            toast.id = Math.random() * Math.pow(10, 16);
+        // prevent duplicate toasts from appearing when multiple UI elements
```
Comment (kzantow) [inline blueocean-core-js/src/js/ToastService.js]: Maybe this service should be called `BestMan`? ;)

File: PULL_REQUEST_TEMPLATE [hunk 2]
```diff
@@ -5,13 +5,13 @@ See [JENKINS-XXXXX](https://issues.jenkins-ci.org/browse/JENKINS-XXXXX).
 # Submitter checklist
 - [ ] Link to JIRA ticket in description, if appropriate.
 - [ ] Change is code complete and matches issue description
-- [ ] Apppropriate unit or acceptance tests or explaination to why this change has no tests
+- [ ] Appropriate unit or acceptance tests or explanation to why this change has no tests
```
Comment (kzantow) [inline PULL_REQUEST_TEMPLATE]: MEGA-ULTRA- 🐝

File: blueocean-core-js/src/js/ToastService.js [hunk 3]
```diff
@@ -25,16 +25,24 @@ export class ToastService {
...
+
+        if (!newToast.id) {
+            newToast.id = Math.random() * Math.pow(10, 16);
         }
 
         // TODO: determine why it's necessary to re-set the "toasts" field to trigger the UI update
-        const copy = this.toasts.slice();
-        copy.push(toast);
-        this.toasts = copy;
+        const toasts = this.toasts.slice();
```
Comment (kzantow) [inline blueocean-core-js/src/js/ToastService.js]: yuck. boo for the need to do this with mobx, understand, but boo

File: blueocean-core-js/src/js/ToastService.js [hunk 4]
```diff
@@ -43,14 +51,31 @@ export class ToastService {
      * @param toast
      */
     removeToast(toast) {
-        this.toasts = this.toasts.filter((item) => {
-            return toast.id !== item.id;
-        });
+        this.toasts = this.toasts.filter((item) =>
```
Comment (kzantow) [inline blueocean-core-js/src/js/ToastService.js]: honestly not sure which of these i prefer... seems to be easier to debug the bracket-enclosed version, would definitely prefer this changed style if it was easier to debug

File: blueocean-core-js/src/js/ToastService.js [hunk 5]
```diff
@@ -43,14 +51,31 @@ export class ToastService {
...
     }
 
+    /**
+     * Returns true if a toast with the same 'text' and 'action' already exists
+     * @param newToast
+     * @returns {boolean}
+     * @private
+     */
+    _hasDuplicate(newToast) {
+        for (const toast of this.toasts) {
```
Comment (kzantow) [inline blueocean-core-js/src/js/ToastService.js]: man, `of` and `in` are pretty easy to overlook in these situations, curses ES6+

File: blueocean-core-js/src/js/ToastService.js (hunk 3 above)
Comment (cliffmeyers) [inline blueocean-core-js/src/js/ToastService.js]: This definitely isn't a straight up MobX limitation: there is tons of sample code on the internet that modify the array in place (MobX actually uses its own ObservableArray under the hood). I just haven't had a chance to roll up sleeves and figure out what the problem is - yet.

File: blueocean-core-js/src/js/UrlBuilder.js [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+    }
+
+    if (!restUrl) {
+        throw new Error('Could not find input URL');
+    }
+
+    return restUrl;
+};
+
+const stripEmptyTokens = (tokens) => {
```
Comment (kzantow) [inline blueocean-core-js/src/js/UrlBuilder.js]: NIT: maybe `trimEmptyTokens`

File: blueocean-core-js/src/js/UrlBuilder.js [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+    }
+
+    if (copy[copy.length - 1] === '') {
+        copy.pop();
+    }
+
+    return copy;
+};
+
+export const isMultiBranchRun = (run) => {
```
Comment (kzantow) [inline blueocean-core-js/src/js/UrlBuilder.js]: this is clever (and awful that it's needed), I needed this function many times ;)

**Comment by michaelneale** [comment]
Functionally seems ok to me - big change though. Would like @kzantow and possibly @scherler to give it a look over. Also can you confirm this runs the ATH (if updated with master) reasonably well locally (either its master branch or ideally: https://github.com/jenkinsci/blueocean-acceptance-test/pull/26 - although that branch probably does pick up known queue-issues).

**Comment by cliffmeyers** [comment]
@michaelneale @i386 could you guys please test the Re-run button the Run Details screen? It appears that it may have regressed after I completed the merge this morning.

**Comment by i386** [comment]
@cliffmeyers I tried this and it worked fine. However, I did find this bug https://issues.jenkins-ci.org/browse/JENKINS-37880 that has nothing todo with your work in this PR. cc @vivek

**Comment by cliffmeyers** [comment]
@i386 ok great! Perhaps I was just running into a dirty node_modules and needed a clean, or maybe that navigation works ok in Safari but not Chrome. Will test more tomorrow.

Regarding the other bug, that might be the elusive race condition with /replay I've seen. If I can add any info to that bug I will.

Thanks again!

**Comment by i386** [comment]
@cliffmeyers always worth a clean - I setup an alias in my bash profile so thats the only way I run maven now :)

**Comment by michaelneale** [comment]
@cliffmeyers I saw an error when running replay from a fav car, may be same thing: https://github.com/jenkinsci/blueocean-plugin/pull/447#issuecomment-243968352 (haven't tried master, and I think it was intermittent).

File: blueocean-core-js/src/js/ToastService.js (hunk 3 above)
Comment (michaelneale) [inline blueocean-core-js/src/js/ToastService.js]: Is the only real reason for this so you could say `toasts.slice()` ? ;) I mean the whole toast service is really a ruse to get to that point in the code? tell me it isn't true.

File: blueocean-core-js/src/js/UrlBuilder.js (hunk 6 above)
Comment (cliffmeyers) [inline blueocean-core-js/src/js/UrlBuilder.js]: Fixed in d751a1a

**Comment by cliffmeyers** [comment]
@i386 @michaelneale the bug was still present in this branch but fixed in c530784.

**Comment by cliffmeyers** [comment]
ATH passed.

`2016-09-01 11:32:14,897 [Thread-3] INFO   (InputStreamHandler.java:47) com.github.eirslett.maven.plugins.frontend.lib.DefaultGulpRunner - OK. 125  total assertions passed. (1m 53s)`

**Comment by cliffmeyers** [comment]
ATH passed locally.

2016-09-01 15:06:17,115 [Thread-3] INFO   (InputStreamHandler.java:47) com.github.eirslett.maven.plugins.frontend.lib.DefaultGulpRunner - OK. 125  total assertions passed. (2m 11s)

File: blueocean-core-js/src/js/components/RunButton.jsx [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+        const name = decodeURIComponent(
+            event.job_ismultibranch ? event.blueocean_job_branch_name : event.blueocean_job_pipeline_name
+        );
+        const runId = event.jenkins_object_id;
+
+        if (event.jenkins_event === 'job_run_started') {
+            const runDetailsUrl = buildRunDetailsUrl(runData);
+
+            toastService.newToast({
+                text: `Started "${name}" #${runId}`,
```
Comment (kzantow) [inline blueocean-core-js/src/js/components/RunButton.jsx]: A lot of duplication between these, seems like might be nice to have a more generalized component that takes a toast message, icon, and action or something.

Comment (kzantow) [inline blueocean-core-js/src/js/components/RunButton.jsx]: I guess there are plenty of differences, too.

**Comment by michaelneale** [comment]
:bee: from me if it all looks good. @kzantow I think cliff wanted you to look over?

**Comment by kzantow** [comment]
@michaelneale yep, reviewing it now, will take a while, it's a large one

File: blueocean-core-js/src/js/components/RunButton.jsx [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+    _onStopClick() {
+        if (this.state.stopping) {
+            return;
+        }
+
+        this.setState({
+            stopping: true,
+        });
+
+        runApi.stopRun(this.props.latestRun);
```
Comment (kzantow) [inline blueocean-core-js/src/js/components/RunButton.jsx]: NIT, maybe just `run`? Could this be used to stop something from the activity, which is a run?

**Comment by kzantow** [comment]
:bee: once conflicts resolved, lots of good changes here :)

File: blueocean-core-js/src/js/components/RunButton.jsx (hunk 9 above)
Comment (cliffmeyers) [inline blueocean-core-js/src/js/components/RunButton.jsx]: I have another PR coming soon behind this one for some refinements and I will make the change there.

**Comment by cliffmeyers** [comment]
ATH passed locally.

`2016-09-02 10:40:38,842 [Thread-3] INFO   (InputStreamHandler.java:47) com.github.eirslett.maven.plugins.frontend.lib.DefaultGulpRunner - OK. 125  total assertions passed. (1m 49s)`

File: blueocean-core-js/src/js/ToastService.js (hunk 3 above)
Comment (scherler) [inline blueocean-core-js/src/js/ToastService.js]: rofl, missing the 

```
// with butter
```

Comment (kzantow) [inline blueocean-core-js/src/js/ToastService.js]: ... butter should be using the spread operator

**Pull request merged**